heron/schedulers/src/java/com/twitter/heron/scheduler/yarn/HeronMasterDriver.java

```
@@ -139,107 +142,129 @@ public HeronMasterDriver(EvaluatorRequestor requestor,
     this.packing = topologyPacking;
     for (Entry<String, ContainerPlan> entry : topologyPacking.containers.entrySet()) {
       Resource reqResource = entry.getValue().resource;
-
       int mem = getMemInMBForExecutor(reqResource);
-      try {
-        launchContainerForExecutor(entry.getKey(), getCpuForExecutor(reqResource), mem);
-      } catch (InterruptedException e) {
-        // Fail deployment of topology if there is a error starting any worker
-        LOG.log(Level.SEVERE, "Container allocation for id:{0} failed. Attempting to close "
-            + "currently allocated resources for {1}", new Object[]{entry.getKey(), topologyName});
-        killTopology();
-        throw new RuntimeException("Failed to allocate container for workers", e);
+      int cores = getCpuForExecutor(reqResource);
+      AllocatedEvaluator container = launchContainerForExecutor(entry.getKey(), cores, mem);
+      if (container == null) {
+        return false;
       }
     }
+    return true;
   }
 
   public void killTopology() {
     LOG.log(Level.INFO, "Kill topology: {0}", topologyName);
-    Map<String, ActiveContext> contexts = clearActiveContexts();
+    isTopologyKilled.set(true);
 
-    for (Entry<String, ActiveContext> entry : contexts.entrySet()) {
-      LOG.log(Level.INFO, "Close context: {0}", entry.getKey());
-      entry.getValue().close();
+    for (HeronWorker worker : multiKeyWorkerMap.getHeronWorkers()) {
+      LOG.log(Level.INFO, "Killing container {0} for executor {1}",
+          new Object[]{worker.evaluator.getId(), worker.workerId});
+      worker.evaluator.close();
     }
   }
 
-  /**
-   * REEF will retain all YARN containers. Currently running heron-executors/REEF-tasks
-   * will be killed and new REEF-tasks representing previously running heron-executor will be
-   * started.
-   */
-  public void restartTopology() {
-    restartContainer(null);
+  public boolean restartTopology() {
```

Are you using the boolean as a substitute for an exception here? If so, we should throw an exception if things that are expected to execute fail to do so.